src/Chapter1.hs

```diff
@@ -541,8 +543,16 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | (x < y && x > z) || (x > y && x < z) = x
+    | (y < x && y > z) || (y > x && y < z) = y
+    | (z < x && z > y) || (z > x && z < y) = z
+    | (x == y) && (y == z) && (x == z) = x
+    | (x == y) = x
+    | (y == z) = y
+    | (x == z) = z
+    | otherwise = 0
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive` if you don't add the `otherwise` case, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you had to add another guard – `| otherwise = 0`, to tell the compiler, that your pattern matching is exhaustive 🙂 However, as you rightly noticed, you indeed covered all cases! So it is guaranteed that it won't come to the last case ever.

Taking that into consideration, you can move any two cases returning the same value under a single otherwise and return, for example, `z` in there.

Also note that a lot of `==` could be moved under other cases, if you use `<=` or `>=` instead of `<` and `>`:

```suggestion
    | (x <= y && x >= z) || (x >= y && x <= z) = x
    | (y <= x && y >= z) || (y >= x && y <= z) = y
    | otherwise = z
```